lib/_tls_wrap.js

```diff
@@ -37,211 +37,243 @@ const tls_wrap = process.binding('tls_wrap');
+  if (owner._newSessionPending) {
+    owner._securePending = true;
+    return;
+  }
+
+  owner._finishInit();
 }
 
 
-function loadSession(self, hello, cb) {
+function loadSession(hello) {
+  const owner = this.owner;
+
   var once = false;
   function onSession(err, session) {
     if (once)
-      return cb(new errors.Error('ERR_MULTIPLE_CALLBACK'));
+      return loadSessionDone(new errors.Error('ERR_MULTIPLE_CALLBACK'), owner);
     once = true;
 
     if (err)
-      return cb(err);
+      return loadSessionDone(err, owner);
 
-    if (!self._handle)
-      return cb(new errors.Error('ERR_SOCKET_CLOSED'));
+    if (owner._handle === null)
+      return loadSessionDone(new errors.Error('ERR_SOCKET_CLOSED'), owner);
 
-    self._handle.loadSession(session);
-    cb(null);
+    owner._handle.loadSession(session);
+    loadSessionDone(null, owner);
   }
 
   if (hello.sessionId.length <= 0 ||
       hello.tlsTicket ||
-      self.server &&
-      !self.server.emit('resumeSession', hello.sessionId, onSession)) {
-    cb(null);
+      owner.server &&
+      !owner.server.emit('resumeSession', hello.sessionId, onSession)) {
+    loadSessionDone(null, owner);
   }
 }
 
+function loadSessionDone(err, self) {
+  if (err !== null)
+    return self.destroy(err);
 
-function loadSNI(self, servername, cb) {
-  if (!servername || !self._SNICallback)
-    return cb(null);
+  self._handle.endParser();
+}
 
-  var once = false;
-  self._SNICallback(servername, function(err, context) {
+
+function loadSNI(info) {
+  const owner = this.owner;
+  const servername = info.servername;
+  if (!servername || !owner._SNICallback)
+    return requestOCSP(null, owner, info);
+
+  let once = false;
+  owner._SNICallback(servername, (err, context) => {
     if (once)
-      return cb(new errors.Error('ERR_MULTIPLE_CALLBACK'));
+      return requestOCSP(new errors.Error('ERR_MULTIPLE_CALLBACK'), owner);
     once = true;
 
     if (err)
-      return cb(err);
+      return requestOCSP(err, owner);
 
-    if (!self._handle)
-      return cb(new errors.Error('ERR_SOCKET_CLOSED'));
+    if (owner._handle === null)
+      return requestOCSP(new errors.Error('ERR_SOCKET_CLOSED'), owner);
 
     // TODO(indutny): eventually disallow raw `SecureContext`
     if (context)
-      self._handle.sni_context = context.context || context;
+      owner._handle.sni_context = context.context || context;
 
-    cb(null, self._handle.sni_context);
+    requestOCSP(null, owner, info);
   });
 }
 
 
-function requestOCSP(self, hello, ctx, cb) {
+function requestOCSP(err, self, hello) {
+  if (err !== null)
+    return self.destroy(err);
+
   if (!hello.OCSPRequest || !self.server)
-    return cb(null);
+    return requestOCSPDone(null, self);
+
+  let ctx = self._handle.sni_context;
 
   if (!ctx)
     ctx = self.server._sharedCreds;
 
   // TLS socket is using a `net.Server` instead of a tls.TLSServer.
   // Some TLS properties like `server._sharedCreds` will not be present
   if (!ctx)
```

Comment:
This could be simplified by moving this in the upper `if (!ctx)` check. This would safe a check in case ctx exists from the beginning on.